### PR TITLE
[luci] Options to preserve input/output in ConvertNCHWToNHWC

### PR DESCRIPTION
**Pull request by jinevening**
This adds options to preserve input/output shape in ConvertNCHWToNHWCPass.

ONE-DCO-1.0-Signed-off-by: Hyukjin Jeong <hj1.jeong@samsung.com>

---
Related to: #5600 
Draft PR: #5601

### CODE REVIEW

File: compiler/luci/pass/include/luci/Pass/ConvertNCHWToNHWCPass.h [hunk 1]
```diff
@@ -35,11 +35,22 @@ namespace luci
...
+
   virtual ~ConvertNCHWToNHWCPass() = default;
 
   const char *name(void) const final { return "luci::ConvertNCHWToNHWCPass"; }
 
   bool run(loco::Graph *g) final;
+
+private:
+  bool _preserve_input;
+  bool _preserve_output;
```
Comment (llFreetimell) [inline compiler/luci/pass/include/luci/Pass/ConvertNCHWToNHWCPass.h:53]:
```suggestion
  bool _preserve_input = false;
  bool _preserve_output = false;
```

or make default constructor `delete` would be better

Comment (jinevening) [reply compiler/luci/pass/include/luci/Pass/ConvertNCHWToNHWCPass.h]: Updated. Thanks.

**Review by llFreetimell — DISMISSED** [review]
LGTM

File: compiler/luci/pass/src/ConvertNCHWToNHWCPass.cpp [hunk 2]
```diff
@@ -513,7 +513,17 @@ bool ConvertNCHWToNHWCPass::run(loco::Graph *g)
     {
       // List of supported Ops
       case luci::CircleOpcode::CIRCLEINPUT:
+        if (!_preserve_input && !has_data_format(node))
```
Comment (seanshpark) [inline compiler/luci/pass/src/ConvertNCHWToNHWCPass.cpp:516]: Q) whay also check `has_data_format` here? and if the input node doesn't have data format why set to `NCHW`?

File: compiler/luci/pass/include/luci/Pass/ConvertNCHWToNHWCPass.h [hunk 3]
```diff
@@ -35,11 +35,22 @@ namespace luci
  */
 struct ConvertNCHWToNHWCPass final : public logo::Pass
 {
+public:
+  ConvertNCHWToNHWCPass(bool preserve_input = false, bool preserve_output = false)
```
Comment (seanshpark) [inline compiler/luci/pass/include/luci/Pass/ConvertNCHWToNHWCPass.h:39]: We usually don't use default argument values. please remove `= false` for `preserve_output`

File: compiler/luci/pass/src/ConvertNCHWToNHWCPass.test.cpp [hunk 4]
```diff
@@ -238,7 +238,7 @@ void check_post_trans(loco::Node *node)
   EXPECT_EQ(2, post_trans_perm->at<loco::DataType::S32>(3));
 }
 
-void run_phase(loco::Graph *g)
+void run_phase(loco::Graph *g, bool preserve_input = false, bool preserve_output = false)
```
Comment (seanshpark) [inline compiler/luci/pass/src/ConvertNCHWToNHWCPass.test.cpp:241]: ditto

File: compiler/luci/pass/src/ConvertNCHWToNHWCPass.cpp (hunk 2 above)
Comment (jinevening) [reply compiler/luci/pass/src/ConvertNCHWToNHWCPass.cpp:516]: ConverNCHWToNHWC works as below.

1. Set data format of target nodes to NCHW (**this should be done once**)
2. Convert the target nodes from NCHW  to NHWC.

The above steps are repeated until no change is detected. To do step 1 once, I check the existence of data_format.

File: compiler/luci/pass/include/luci/Pass/ConvertNCHWToNHWCPass.h (hunk 3 above)
Comment (jinevening) [reply compiler/luci/pass/include/luci/Pass/ConvertNCHWToNHWCPass.h:39]: Can I ask you why we don't use default arguments?

Default arguments can reduce the changes in the test codes. I can remove them in this PR, but am curious about the reason.

File: compiler/luci/pass/src/ConvertNCHWToNHWCPass.cpp (hunk 2 above)
Comment (seanshpark) [reply compiler/luci/pass/src/ConvertNCHWToNHWCPass.cpp:516]: Thanks! I've forgot -_-;

File: compiler/luci/pass/include/luci/Pass/ConvertNCHWToNHWCPass.h (hunk 3 above)
Comment (seanshpark) [reply compiler/luci/pass/include/luci/Pass/ConvertNCHWToNHWCPass.h:39]: Cause of code is implicit and implicit codes may make problems when there are more changes later. this was from when @parjong started.

Comment (jinevening) [reply compiler/luci/pass/include/luci/Pass/ConvertNCHWToNHWCPass.h:39]: I see. Thanks :)

**Review by mhs4670go — APPROVED** [review]
LGTM

**Review by llFreetimell — APPROVED** [review]
LGTM

**Review by seanshpark — APPROVED** [review]
LGTM thank you!

**Pull request merged**